dali/driver/hid.py

```diff
@@ -207,14 +207,22 @@ async def send(self, command, in_transaction=False, exceptions=None):
             if not in_transaction:
                 self.transaction_lock.release()
 
-    async def run_sequence(self, seq, progress=None):
+    async def run_sequence(self, cmd_seq, progress=None):
+        """Run a command sequence as a transaction
+        """
+        if hasattr(cmd_seq, 'run'):
+            seq = cmd_seq.run()
+        else:
+            seq = cmd_seq
```

Comment:
I want to allow plain generators to be passed in; this mechanism is useful as the way to specify a series of commands are to be run as a DALI transaction as well as to run pre-packaged sequences.

I should add an optional argument to `run_sequence` for the transaction priority, assuming we can find some interface hardware that implements it!

Reply:
Unfortunatley, I don't have access to the most recent DALI specifications, so I don't quite get what DALI transactions are supposed to be.

Another solution would be a type of sequence that wraps a list of commands:
```python
class ListOfCommands(Sequence):

    def __init__(self, commands):
        self.commands = commands

    def run(self):
        for cmd in self.commands:
            yield cmd
```

Reply:
Transactions are a concept of python-dali to enable complex interactions with devices while being able to access the bus from multiple tasks concurrently, correct?

Reply:
They are that, but they are also the same concept implemented at the DALI bus level for multiple bus masters on a single bus.

IEC-62386 part 101 section 8.3 defines five "frame priorities" in terms of the bus settling time before a forward frame can be transmitted:
* Priority 1: 13.5–14.7ms
* Priority 2: 14.9–16.1ms
* Priority 3: 16.3–17.7ms
* Priority 4: 17.9–19.3ms
* Priority 5: 19.5–21.1ms

The bus master is expected to pick a random time between the minimum and maximum for the relevant priority each time it needs to transmit a frame.

A "transaction" is a frame at one of priorities 2–5, followed by zero or more frames at priority 1.  Transaction priorities are allocated as follows:
* Priority 2: user instigated actions for switching or dimming the lights
* Priority 3: configuration of a bus unit and for those event messages that are not covered by Priorities 2 and 4
* Priority 4: automatic actions for switching or dimming the lights
* Priority 5: periodic query commands

A number of frame sequences are defined to be sent as transactions: any "send-twice" command, QueryDeviceTypes, reading and writing memory, commissioning, etc.

python-dali doesn't have API support for this at the moment because none of us has multi-master capable hardware (or if we do, we don't know how to drive it properly)!